test/e2e/storage/csi_volumes.go

```diff
@@ -0,0 +1,242 @@
+			Name: config.Prefix + "-role-binding",
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      sa.GetName(),
+				Namespace: sa.GetNamespace(),
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			Kind:     "ClusterRole",
+			Name:     clusterRole.GetName(),
+			APIGroup: "rbac.authorization.k8s.io",
+		},
+	}
+
+	clusterRoleBindingClient.Delete(binding.GetName(), &metav1.DeleteOptions{})
+	err := wait.Poll(2*time.Second, 10*time.Minute, func() (bool, error) {
+		_, err := clusterRoleBindingClient.Get(binding.GetName(), metav1.GetOptions{})
+		return apierrs.IsNotFound(err), nil
+	})
+	framework.ExpectNoError(err, "Timed out waiting for deletion: %v", err)
+
+	if teardown {
+		return nil
+	}
+
+	ret, err := clusterRoleBindingClient.Create(binding)
+	if err != nil {
+		framework.ExpectNoError(err, "Failed to create %s role binding: %v", binding.GetName(), err)
+	}
+
+	return ret
+}
+
+var _ = SIGDescribe("CSI Volumes [Feature:CSI]", func() {
+	f := framework.NewDefaultFramework("csi-mock-plugin")
+
+	var (
+		cs     clientset.Interface
+		ns     *v1.Namespace
+		node   v1.Node
+		config framework.VolumeTestConfig
+		suffix string
+	)
+
+	BeforeEach(func() {
+		cs = f.ClientSet
+		ns = f.Namespace
+		nodes := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
+		node = nodes.Items[rand.Intn(len(nodes.Items))]
+		config = framework.VolumeTestConfig{
+			Namespace:         ns.Name,
+			Prefix:            "csi",
+			ClientNodeName:    node.Name,
+			ServerNodeName:    node.Name,
+			WaitForCompletion: true,
+		}
+		suffix = ns.Name
+	})
+
+	// Create one of these for each of the drivers to be tested
+	// CSI hostPath driver test
+	Describe("Sanity CSI plugin test using hostPath CSI driver", func() {
+
+		var (
+			clusterRole    *rbacv1.ClusterRole
+			serviceAccount *v1.ServiceAccount
+		)
+
+		BeforeEach(func() {
+			By("deploying csi hostpath driver")
+			clusterRole = externalAttacherClusterRole(cs, config, false)
+			serviceAccount = externalAttacherServiceAccout(cs, config, false)
+			externalAttacherClusterRoleBinding(cs, config, false, serviceAccount, clusterRole)
+			csiHostPathPod(cs, config, false, f, serviceAccount)
+		})
+
+		AfterEach(func() {
+			By("uninstalling csi hostpath driver")
+			csiHostPathPod(cs, config, true, f, serviceAccount)
+			externalAttacherClusterRoleBinding(cs, config, true, serviceAccount, clusterRole)
+			serviceAccount = externalAttacherServiceAccout(cs, config, true)
+			clusterRole = externalAttacherClusterRole(cs, config, true)
+		})
+
+		It("should provision storage with a hostPath CSI driver", func() {
+			t := storageClassTest{
+				name:         "csi-hostpath",
+				provisioner:  "csi-hostpath",
+				parameters:   map[string]string{},
+				claimSize:    "1Gi",
+				expectedSize: "1Gi",
+				nodeName:     node.Name,
+			}
+
+			claim := newClaim(t, ns.GetName(), "")
+			class := newStorageClass(t, ns.GetName(), "")
+			testDynamicProvisioning(t, cs, claim, class)
```

Comment:
Any plans to also test mount/unmount, attach/detach?

Reply:
Right now, we would need a driver that supports attach/detach. But mount/unmount is tested by `testDynamicProvisioning()`.

Reply:
Actually, now that I think about it, the test does test it. It all depends on the driver, but the e2e test does send out all the CSI calls to attach/detach and mount/unmount. All the CSI driver we have right now do attach/deatch and mount/unmount on the `NodePublish/Unpublish` calls done from the kubelet to the CSI driver. Probably when we get CSI drivers for the cloud vendors, the test would exercise attach/detach done from the `external-attacher` side car container to the CSI driver, then mount/unmount would be done by the kubelet to the CSI driver.

So, tldr: Yes, both are tested.

Reply:
Now that I think about it, maybe I should change the text on line 227 to say something like: `should provision, attach/detach, then mount/unmount storage with a hostPath CSI driver`. Would that make sense?